### PR TITLE
FEAT(client): Toggle positional audio shortcut #6133

### PR DESCRIPTION
**Pull request by GrossTrevor**
Constructed framework of a new shortcut to toggle positional audio, but do not know what functions to use or where to go from here. This is a work in progress that I would need some help with. Forgive me, I'm afraid this is my first open-source commit like this. This is concerning issue #6133 to which I am assigned.

- Added new Global Shortcut type in GlobalShortcutTypes.h
- Remade new assertion in same file
- Initialized shortcut as a new GlobalShortcut in MainWindow.cpp
- Created a new function that corresponds to the new shortcut (no real code in there yet, this is where I need help)
- Associated definitions in MainWindow.h

This is for a school assignment due Friday, February 9th, so please provide some guidance as soon as possible if you can. Thanks!

### CODE REVIEW

File: src/mumble/MainWindow.cpp [hunk 1]
```diff
@@ -3350,6 +3355,15 @@ void MainWindow::on_gsHelpVersionCheck_triggered(bool down, QVariant) {
 	versionCheck();
 }
 
+void MainWindow::on_gsTogglePositionalAudio_triggered(bool down, QVariant) {
+	if (!down) {
+		return;
+	}
+
+	// WIP: Toggle positional audio
+	Global::get().pluginManager->togglePositionalAudio();
```
Comment (Hartmnt) [inline src/mumble/MainWindow.cpp:3364]: I assume this is the part where you are unsure what to do. If I understand the original feature request correctly, we simply want a shortcut that toggles positional audio on a global level. That should mean it will be enough to toggle the global positional audio setting. (the same as toggling the positional audio checkbox in the respective settings page)

So lets look at the AudioOutput settings page UI definition:

https://github.com/mumble-voip/mumble/blob/c4b5858d141f76cce553be2f74dfc4291989fc9b/src/mumble/AudioOutput.ui#L489-L495

There we can see that the checkbox is named ``qcbPositional``

We can now scan the code for ``qcbPositional`` and find the respective Mumble setting here:

https://github.com/mumble-voip/mumble/blob/c4b5858d141f76cce553be2f74dfc4291989fc9b/src/mumble/AudioConfigDialog.cpp#L766

Which means we are interested in toggling ``bPositionalAudio``.

So the next step is to figure out how to do that. Luckily, other shortcuts exists which already toggle some settings. Lets look at "Toggle Text-To-Speech":

https://github.com/mumble-voip/mumble/blob/7bed82e8c2aba2d4405861673e2af688e1a5b819/src/mumble/MainWindow.cpp#L3326-L3332
https://github.com/mumble-voip/mumble/blob/7bed82e8c2aba2d4405861673e2af688e1a5b819/src/mumble/MainWindow.cpp#L4152-L4154

So I think it should be enough to follow the same steps as the ``Toggle Text-To-Speech`` shortcut, but with ``bPositionalAudio`` instead of ``bTTS``.

**Review by Hartmnt — COMMENTED** [review]
Otherwise, your merge request looks good so far. In the end you will need to add a translation commit. But first you just need to worry about implementing the logic :)

File: src/mumble/MainWindow.cpp (hunk 1 above)
Comment (GrossTrevor) [reply src/mumble/MainWindow.cpp:3364]: Awesome, thank you so much! I'll get on that right away.

**Comment by GrossTrevor** [comment]
I added another commit that mimics the function on_gsAudioTTS_triggered(bool down, QVariant) as suggested. Some of the checks are still failing and, as a novice, I really have no idea what that means for my commit. Are there errors in my addition that need to be fixed?

**Comment by Hartmnt** [comment]
> I added another commit that mimics the function on_gsAudioTTS_triggered(bool down, QVariant) as suggested. Some of the checks are still failing and, as a novice, I really have no idea what that means for my commit. Are there errors in my addition that need to be fixed?

Please squash your two existing commits into one. While you are at it you can also include a ``Fixes #6133`` to the commit message.
And then add a translation commit using the Python script ``updatetranslations.py`` in our repo. That script will automatically add a translation commit to your branch without further user interaction. After that you can update this merge request by force pushing your branch ``git push origin positional-audio-shortcut --force-with-lease``. The force-push is necessary, because of the squashed commits.

**Comment by GrossTrevor** [comment]
> > I added another commit that mimics the function on_gsAudioTTS_triggered(bool down, QVariant) as suggested. Some of the checks are still failing and, as a novice, I really have no idea what that means for my commit. Are there errors in my addition that need to be fixed?
> 
> Please squash your two existing commits into one. While you are at it you can also include a `Fixes #6133` in the commit message. And then add a translation commit using the Python script `updatetranslations.py` in our repo. That script will automatically add a translation commit to your branch without further user interaction. After that you can update this merge request by force pushing your branch `git push origin positional-audio-shortcut --force-with-lease`. The force-push is necessary, because of the squashed commits.

I've been messing with this for a while now and I cannot get the script to run because it cannot find Lupdate. I have QT for open source on Windows but I have never used the program before. I already squished the commits. Is there a way I can perform the function of `updatetranslations.py` manually?

**Comment by Hartmnt** [comment]
> I've been messing with this for a while now and I cannot get the script to run because it cannot find Lupdate. I have QT for open source on Windows but I have never used the program before. I already squished the commits. Is there a way I can perform the function of `updatetranslations.py` manually?

Not sure what the issue is, but I certainly have no idea how to get this to work on windows. No problem though, I will gladly add the translation commit for you. You just need to update your branch to include the single squashed commit :)

**Comment by Hartmnt** [comment]
One line of your code change was not following the clang-10 format:

```
diff --git a/src/mumble/MainWindow.cpp b/src/mumble/MainWindow.cpp
index 9c8d8479c..dae803de1 100644
--- a/src/mumble/MainWindow.cpp
+++ b/src/mumble/MainWindow.cpp
@@ -412,8 +412,8 @@ void MainWindow::createActions() {
        gsHelpVersionCheck->setObjectName(QLatin1String("gsHelpVersionCheck"));
        gsHelpVersionCheck->qsWhatsThis = tr("This will check if mumble is up to date");
 
-       gsTogglePositionalAudio =
-               new GlobalShortcut(this, GlobalShortcutType::TogglePositionalAudio, tr("Toggle positional audio", "Global Shortcut"));
+       gsTogglePositionalAudio = new GlobalShortcut(this, GlobalShortcutType::TogglePositionalAudio,
+                                                                                                tr("Toggle positional audio", "Global Shortcut"));
        gsTogglePositionalAudio->setObjectName(QLatin1String("gsTogglePositionalAudio"));
        gsTogglePositionalAudio->qsWhatsThis = tr("This will turn on/off positional audio");
```

For the test to pass, that needs to be changed as well. I pushed the translation commit already. Is your git-fu strong enough to edit the code commit without touching the translation commit :sweat_smile: ?
If not, just do the change and force push the branch without my translation commit. I can always re-add it.

**Comment by GrossTrevor** [comment]
> One line of your code change was not following the clang-10 format:
> 
> ```
> diff --git a/src/mumble/MainWindow.cpp b/src/mumble/MainWindow.cpp
> index 9c8d8479c..dae803de1 100644
> --- a/src/mumble/MainWindow.cpp
> +++ b/src/mumble/MainWindow.cpp
> @@ -412,8 +412,8 @@ void MainWindow::createActions() {
>         gsHelpVersionCheck->setObjectName(QLatin1String("gsHelpVersionCheck"));
>         gsHelpVersionCheck->qsWhatsThis = tr("This will check if mumble is up to date");
>  
> -       gsTogglePositionalAudio =
> -               new GlobalShortcut(this, GlobalShortcutType::TogglePositionalAudio, tr("Toggle positional audio", "Global Shortcut"));
> +       gsTogglePositionalAudio = new GlobalShortcut(this, GlobalShortcutType::TogglePositionalAudio,
> +                                                                                                tr("Toggle positional audio", "Global Shortcut"));
>         gsTogglePositionalAudio->setObjectName(QLatin1String("gsTogglePositionalAudio"));
>         gsTogglePositionalAudio->qsWhatsThis = tr("This will turn on/off positional audio");
> ```
> 
> For the test to pass, that needs to be changed as well. I pushed the translation commit already. Is your git-fu strong enough to edit the code commit without touching the translation commit 😅 ? If not, just do the change and force push the branch without my translation commit. I can always re-add it.

Huh that is weird. I copied and pasted the above shortcut initializations to ensure my formatting was the exact same, but I guess that did not work. It also looks correct on my machine, but who knows. Are the lines with pluses what I should change it to? Also, I can guarantee you that I will find a way to mess up your translation commit, but I will try lol.

**Comment by Hartmnt** [comment]
> Huh that is weird. I copied and pasted the above shortcut initializations to ensure my formatting was the exact same, but I guess that did not work. It also looks correct on my machine, but who knows. Are the lines with pluses what I should change it to? Also, I can guarantee you that I will find a way to mess up your translation commit, but I will try lol.

The snippet above is a diff. Your lines (the ones with the ``-``) are to be removed and the newly formatted ones (with the ``+``) are to be added. Your code is functional as is, but we require it to be formatted according to the clang-10 specification.

For this line in particular that just means it is not happy with the line length. Since the name of the variable changed, it does not want it to be formatted as the shortcut above. There is a tool to automate this (``clang-format``), but I have no idea, if and how to use that on Windows ;) For this single line it is probably easier to just change it by hand.

If you follow the link of the failing test here on GitHub https://github.com/mumble-voip/mumble/actions/runs/7775343666/job/21201217927?pr=6326 you can also see where the test complains about the line.

**Comment by GrossTrevor** [comment]
> > Huh that is weird. I copied and pasted the above shortcut initializations to ensure my formatting was the exact same, but I guess that did not work. It also looks correct on my machine, but who knows. Are the lines with pluses what I should change it to? Also, I can guarantee you that I will find a way to mess up your translation commit, but I will try lol.
> 
> The snippet above is a diff. Your lines (the ones with the `-`) are to be removed and the newly formatted ones (with the `+`) are to be added. Your code is functional as is, but we require it to be formatted according to the clang-10 specification.
> 
> For this line in particular that just means it is not happy with the line length. Since the name of the variable changed, it does not want it to be formatted as the shortcut above. There is a tool to automate this (`clang-format`), but I have no idea, if and how to use that on Windows ;) For this single line it is probably easier to just change it by hand.
> 
> If you follow the link of the failing test here on GitHub https://github.com/mumble-voip/mumble/actions/runs/7775343666/job/21201217927?pr=6326 you can also see where the test complains about the line.

Just pushed it again. This is what the code looks like on my machine (hopefully it is correct):
image

**Comment by Hartmnt** [comment]
> Just pushed it again. This is what the code looks like on my machine (hopefully it is correct):

Sometimes a project agrees to ditch subjective code beauty for pragmatic auto-formatting :) Mumble is such a project

**Review by Hartmnt — APPROVED** [review]
Compiled and tested*, LGTM

* The positional audio setting is correctly toggled. I was not able to test it with the manual link plugin in a real world scenario, because that is currently broken on master

File: src/mumble/MainWindow.cpp [hunk 2]
```diff
@@ -412,6 +412,11 @@ void MainWindow::createActions() {
 	gsHelpVersionCheck->setObjectName(QLatin1String("gsHelpVersionCheck"));
 	gsHelpVersionCheck->qsWhatsThis = tr("This will check if mumble is up to date");
 
+	gsTogglePositionalAudio = new GlobalShortcut(this, GlobalShortcutType::TogglePositionalAudio,
+												 tr("Toggle positional audio", "Global Shortcut"));
+	gsTogglePositionalAudio->setObjectName(QLatin1String("gsTogglePositionalAudio"));
+	gsTogglePositionalAudio->qsWhatsThis = tr("This will turn on/off positional audio");
```
Comment (Krzmbrzl) [inline src/mumble/MainWindow.cpp:418]:
```suggestion
	gsTogglePositionalAudio->qsWhatsThis = tr("This will toggle positional audio on/off");
```

File: src/mumble/MainWindow.h [hunk 3]
```diff
@@ -431,6 +433,7 @@ public slots:
 	void openAboutDialog();
 	void openAboutQtDialog();
 	void versionCheck();
+	void togglePositionalAudio(const bool &newState);
```
Comment (Krzmbrzl) [inline src/mumble/MainWindow.h:436]: Why do you take the boolean parameter by reference? This seems unnecessary as the size of a boolean is quite a bit smaller than the size of a pointer (which is what a reference really is, under the hood) and therefore passing the boolean by value should be more efficient.
Premature optimizations aside, I think it would also be clearer to pass by value as passing a boolean by reference introduces the semantics that it shall act as some kind of out-parameter (though that is of course not possible here due to the `const` qualification).

So in short: Unless you have good reason to do so, I would prefer if we passed the boolean by value instead of by reference.

File: src/mumble/MainWindow.cpp [hunk 4]
```diff
@@ -4152,6 +4165,10 @@ void MainWindow::versionCheck() {
 	new VersionCheck(false, this);
 }
 
+void MainWindow::togglePositionalAudio(const bool &newState) {
+	Global::get().s.bPositionalAudio = newState;
```
Comment (Krzmbrzl) [inline src/mumble/MainWindow.cpp:4169]: This does not what the name suggests. If it is supposed to _toggle_ the positional audio state, this function shouldn't take a parameter for the new state as the new state is simply the inverted old state (e.g. `state = !state`). If in fact you want to be able to _set_ the state of positional audio, the function should be called `setPositionalAudio` or even better `enablePositionalAudio`.

**Review by Krzmbrzl — CHANGES_REQUESTED** [review]

**Comment by Krzmbrzl** [comment]
I just saw that the odd naming of toggle-functions is also present in existing functions. Guess that slipped under my radar. Nonetheless, we should strive to have new code match high standards :)

**Comment by Hartmnt** [comment]
@GrossTrevor Hey, I know it is past your assignment due date, but would you be able to apply the requested changes? We can do it for you, if you do not have the time or motivation for it, but your commits might lose some authorship information. Maybe you need the proof that your commit was actually merged or something :sweat_smile: 

Anyway, the other existing methods have already been refactored on master ( https://github.com/mumble-voip/mumble/commit/9b9d88c9c6c8ac4d3fba4f7a2739e935ed33dfd3 ) so you can see exactly how it is now supposed to look :)

**Review by Krzmbrzl — APPROVED** [review]

**Comment by Krzmbrzl** [comment]
Thank you for having worked on this :1:
I have fixed the remaining issues myself in order for this to be merged.
